oxidation/libparsec/crates/crypto/src/common/sequester.rs

```diff
@@ -0,0 +1,224 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+use crate::{CryptoError, CryptoResult};
+
+pub(crate) trait EnforceSerialize {
+    const ALGORITHM: &'static [u8];
+    const SIZE_IN_BITS: usize;
+    const SIZE_IN_BYTES: usize = Self::SIZE_IN_BITS / 8;
+
+    /// Here we avoid uncessary allocation & enforce output has `key_size`
+    fn serialize(&self, output: &[u8], data: &[u8]) -> Vec<u8> {
+        assert!(output.len() <= Self::SIZE_IN_BYTES);
+        let mut res = vec![0; Self::ALGORITHM.len() + 1 + Self::SIZE_IN_BYTES + data.len()];
+
+        let (algorithm_part, others) = res.split_at_mut(Self::ALGORITHM.len());
+        let (colon, others) = others.split_at_mut(1);
+        // Here we enforce output has key size with zeros
+        let (_zeros, others) = others.split_at_mut(Self::SIZE_IN_BYTES - output.len());
+        let (output_part, data_part) = others.split_at_mut(output.len());
+
+        algorithm_part.copy_from_slice(Self::ALGORITHM);
+        colon[0] = b':';
+        output_part.copy_from_slice(output);
+        data_part.copy_from_slice(data);
+
+        res
+    }
+}
+
+pub(crate) trait EnforceDeserialize {
+    const ALGORITHM: &'static [u8];
+    const SIZE_IN_BITS: usize;
+    const SIZE_IN_BYTES: usize = Self::SIZE_IN_BITS / 8;
+
+    fn deserialize(data: &[u8]) -> CryptoResult<(&[u8], &[u8])> {
+        let index = data
+            .iter()
+            .position(|x| *x == b':')
+            .ok_or(CryptoError::Decryption)?;
+        let (algo, output_and_data) = data.split_at(index + 1);
+
+        if &algo[..index] != Self::ALGORITHM {
+            return Err(CryptoError::Algorithm(
+                String::from_utf8_lossy(&algo[..index]).into(),
+            ));
+        }
+
+        Ok(output_and_data.split_at(Self::SIZE_IN_BYTES))
+    }
+}
+
+#[test]
```

Comment:
move the tests to `libparsec/crates/crypto/tests` for consistency

(besides, this test is not in the `#[cfg(test)]` block)

Reply:
On top of that there is not enough test considering how much stuff is implemented

For instance test on serialization is lacking, and also to tests on error handling (typically providing invalid data to decrypt/verify)

Reply:
It was due to the `cfg(test)`, `crypto/tests` are integration test, so this is not compiled in `cfg(test)`, but if we can generate any size, this can be moved

Reply:
Moreover #[test] doesn't need to be in `cfg(test)` block

Reply:
Done